File: shared/types/training.ts
export type AbsenceReason = "injury" | "personal" | "sanction" | "other";

export const ABSENCE_REASON_LABELS: Record<AbsenceReason, string> = {
  injury: "Lesión",
  personal: "Personal",
  sanction: "Sanción",
  other: "Otro",
};

export interface TrainingDrillGroupPlayer {
  id: number;
  first_name: string;
  last_name: string;
}

export interface TrainingDrillGroup {
  id: number;
  training_drill_id: number;
  group_number: number;
  players: TrainingDrillGroupPlayer[];
}

export interface TrainingDrillGroupUpsert {
  groups: Array<{
    group_number: number;
    player_ids: number[];
  }>;
}

export interface TrainingDrill {
  id: number;
  training_id: number;
  drill_id: number;
  position: number;
  notes: string | null;
  duration_minutes: number | null;
  drill_title: string | null;
  drill_type: string | null;
  groups: TrainingDrillGroup[];
}

export interface TrainingAttendance {
  id: number;
  training_id: number;
  player_id: number;
  attended: boolean;
  is_late: boolean;
  absence_reason: AbsenceReason | null;
  notes: string | null;
  player_first_name: string | null;
  player_last_name: string | null;
}

export interface Training {
  id: number;
  team_id: number;
  season_id: number;
  date: string;
  title: string;
  notes: string | null;
  created_by: number | null;
  created_at: string;
  archived_at: string | null;
  training_drills: TrainingDrill[];
  training_attendances: TrainingAttendance[];
}

export interface TrainingCreate {
  title: string;
  date: string;
  season_id: number;
  notes?: string | null;
}

export interface TrainingUpdate {
  title?: string;
  date?: string;
  notes?: string | null;
}

export interface TrainingDrillAdd {
  drill_id: number;
  notes?: string | null;
  duration_minutes?: number | null;
}

export interface TrainingDrillUpdate {
  duration_minutes?: number | null;
  notes?: string | null;
}

export interface TrainingDrillReorderItem {
  drill_id: number;
  position: number;
}

export interface AttendanceUpdate {
  player_id: number;
  attended: boolean;
  is_late?: boolean;
  absence_reason?: AbsenceReason | null;
  notes?: string | null;
}

export interface TrainingBulkDrillItem {
  drill_id: number;
  duration_minutes?: number | null;
}

export interface TrainingBulkItem {
  title: string;
  date: string;
  notes?: string | null;
  drills: TrainingBulkDrillItem[];
}

export interface TrainingBulkCreate {
  season_id: number;
  trainings: TrainingBulkItem[];
}

File: backend/app/schemas/training.py
from datetime import datetime

from pydantic import BaseModel, model_validator

from app.models.training import AbsenceReason


class TrainingDrillGroupPlayerResponse(BaseModel):
    model_config = {"from_attributes": True}
    player_id: int


class TrainingDrillGroupResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    group_number: int
    players: list[TrainingDrillGroupPlayerResponse] = []


class TrainingDrillResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    training_id: int
    drill_id: int
    position: int
    notes: str | None
    duration_minutes: int | None = None
    drill_title: str | None = None
    drill_type: str | None = None
    groups: list[TrainingDrillGroupResponse] = []


class TrainingAttendanceResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    training_id: int
    player_id: int
    attended: bool
    is_late: bool = False
    absence_reason: AbsenceReason | None = None
    notes: str | None = None
    player_first_name: str | None = None
    player_last_name: str | None = None


class TrainingResponse(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    team_id: int
    season_id: int
    date: datetime
    title: str
    notes: str | None
    created_by: int | None
    created_at: datetime
    archived_at: datetime | None
    training_drills: list[TrainingDrillResponse] = []
    training_attendances: list[TrainingAttendanceResponse] = []


class TrainingCreate(BaseModel):
    title: str
    date: datetime
    season_id: int
    notes: str | None = None


class TrainingUpdate(BaseModel):
    title: str | None = None
    date: datetime | None = None
    notes: str | None = None


class TrainingDrillAdd(BaseModel):
    drill_id: int
    duration_minutes: int | None = None
    notes: str | None = None


class TrainingDrillUpdate(BaseModel):
    duration_minutes: int | None = None
    notes: str | None = None


class TrainingDrillReorderItem(BaseModel):
    drill_id: int
    position: int


class AttendanceUpdate(BaseModel):
    player_id: int
    attended: bool
    is_late: bool = False
    absence_reason: AbsenceReason | None = None
    notes: str | None = None

    @model_validator(mode="after")
    def check_consistency(self) -> "AttendanceUpdate":
        if self.attended:
            if self.absence_reason is not None:
                raise ValueError("absence_reason debe ser null si attended=True")
        else:
            if self.absence_reason is None:
                raise ValueError("absence_reason es obligatorio si attended=False")
            if self.is_late:
                raise ValueError("is_late debe ser false si attended=False")
        return self


class TrainingDrillGroupUpsert(BaseModel):
    class GroupItem(BaseModel):
        group_number: int
        player_ids: list[int]
    groups: list[GroupItem]


class TrainingBulkDrillItem(BaseModel):
    drill_id: int
    duration_minutes: int | None = None


class TrainingBulkItem(BaseModel):
    title: str
    date: datetime
    notes: str | None = None
    drills: list[TrainingBulkDrillItem] = []


class TrainingBulkCreate(BaseModel):
    season_id: int
    trainings: list[TrainingBulkItem]
